### ISSUE TITLE
Redis `fastify-rate-limit-*` key is deleted automatically after 1 minute despite timeWindow being set to 5 minutes

### ISSUE THREAD

**Issue by kruczyna** [fastify/fastify-rate-limit#175]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure it has not already been reported

### Fastify version

3.14.2

### Plugin version

5.6.0

### Node.js version

14.16.0

### Operating system

macOS

### Operating system version (i.e. 20.04, 11.3, 10)

11.4

### Description

Redis `fastify-rate-limit-*` key is deleted automatically after 1 minute even if the timeWindow is set to '5 minutes'. I've tried with different timeWindows, both in string and number notation. 
I do not have anywhere a TTL set for the keys in Redis itself, so AFAIK they should be stored until the timeWindow reaches it's limit. 
The issue occurs when I'm connected to Redis both on GCP and locally, something seems to be dumping the key generator from keyGenerator, but I really do not have any clue by now what that might be. 
Could anyone help me with that? 
It seems like a bug, but I am quite new in the field, so it's possible it's a rookie mistake made somewhere. 

### Steps to Reproduce

A. Register FastifyRateLimit on the global server configuration along with local Redis connection. Set global to false, so you're able to create custom `keyGenerator` functions for each route: 

```
server.register(FastifyRateLimit, {
    global: false,
    redis: new Redis({ host: '127.0.0.1', port: 6379})
... })
```
B. Mount `rateLimit` function on one of your routes (in my case a POST one). Return hardcoded 45 for easier reproduction
+ check out if a new key `fastify-rate-limit-45` appeared, via: `get fastify-rate-limit-45`

```
preHandler: [
        fastify.rateLimit({
          max: 1,
          timeWindow: '5 minutes',
          keyGenerator: function (request) {
            return 45
          }
```
C. Spin up the local env and hit the endpoint twice to trigger the rate-limit block
D. Wait for 1 minute
E. Hit the same endpoint again 

**Actual result**:
-> The hit is not blocked, although it should since the timeWindow is set to 5 minutes
-> When executing `get fastify-rate-limit-45` in telnet you'll get a `-1` response - the key has been deleted

### Expected Behavior

- the hit is blocked when timeWindow is not over
- key is not dumped after 1 minutes

**Comment by mcollina** [comment]
Can you please write an example that reproduce your problem?

It might be that this logic (https://github.com/fastify/fastify-rate-limit/blob/master/store/RedisStore.js#L13-L24) is wrong. In case would you like to send a PR?

**Comment by kruczyna** [comment]
@mcollina I've iterated back and ended up having a global config for all the routes in server.ts file. The "x-ratelimit-reset" header is then set properly to the value given in `timeWindow` (in this case 300):
```
  server.register(FastifyRateLimit, {
    global: true,
    max: 1,
    timeWindow: '5 minutes',
    redis: new Redis({ host: '127.0.0.1', port: 6379 }),
    keyGenerator: function (request) {
      return 45
    },
    errorResponseBuilder: function(req, res) {
      return {
        code: 429,
        error: 'Too Many Requests',
        message: `I only allow ${res.max} requests per ${res.after}. Try again soon.`,
        date: 'Request made at: ' + new Date().toISOString(),
        }
      },
    })
```
Although, I wanted to have a separate config for each route and I need the value which is hidden in the response body. I can see that specs say to put it in a config object, but that's not doing the job for me since the body is not being parsed yet at the time rateLimit is executed and the body is `null`. 
To override this I've put the `fastify.rateLimit` into the preHandler function and that's when the "x-ratelimit-reset" is automatically set to 60, no matter what I'm giving as `timeWindow` value.
So I guess, since I was trying to override, this is not a bug case, right? 🙈 
Either way would anyone have any suggestions with what to do with such case? Is my only option exposing my targeted key from the body as hashed value in the header so I can reach it with the lib? 🤔

**Comment by kruczyna** [comment]
@Eomm Thank you for the hints, I've applied them, but the 60 seconds timeout for `x-ratelimit-remaining` is still set buy default 🤷 
I did not use the `config` object, because the body of the request is not getting parsed then and returns `null`. The body parsing happens before the [preValidation](https://www.fastify.io/docs/latest/Hooks/#onRequest) hook, so I do not have yet access to it if I set the config object. 
Btw maybe we should add an info about it to the README.md within `Options on the endpoint itself` section? 🤔

**Comment by Eomm** [comment]
I agree that more `### titles` would help users to understand better the boundaries!

> I did not use the config object, because

Right! I forgot that this plugin adds an `onRequest` hook.

> I've applied them, but the 60 seconds timeout for x-ratelimit-remaining is still set

Would you like to arrange a PR with a failing test to deepen your issue?

**Comment by kruczyna** [comment]
Sure, with pleasure! 
I've done a quick setup to sit on it later, but I can see that already one test is failing: 
```
 FAIL  test/global-rate-limit.test.js
 ✖ should be equal

      t.equal(res.headers['x-ratelimit-remaining'], 0)
      t.equal(res.headers['x-ratelimit-reset'], 0)
--------^

      fastify.inject('/', (err, res) => {

  test: test/global-rate-limit.test.js With redis store
  found: 1
  wanted: 0
  compare: ===
  at:
    line: 251
    column: 9
    file: test/global-rate-limit.test.js
  stack: |
    test/global-rate-limit.test.js:251:9
    node_modules/light-my-request/lib/response.js:31:28
```
Could be that's the source?

**Comment by Eomm** [comment]
The CI is green and it run 7 hours ago https://github.com/fastify/fastify-rate-limit/actions/runs/1141573566

What redis version are you using?

**Comment by kruczyna** [comment]
I have it in: `redis_version:6.2.5`. I've installed all packages via `npm install` after cloning the repo and runner `npm test`. Is there some more configuration to make before running the tests that I am not aware of? 🤔 
Either way, I have to postpone work on this investigation until Monday since I am going off for vacation. Just wanted to let you know that I did not get AWOL on you and I'll be back next week 🤘

**Comment by Eomm** [comment]
> there some more configuration to make before running the tests that I am not aware of? 🤔

No, there is no other configuration to apply

**Comment by Eomm** [comment]
I thought that you could add a console log in your `node_module/fastify-redis` installation to debug your behaviour:

https://github.com/fastify/fastify-rate-limit/blob/8290c305990c12714e2c584a42336be851fbfd6a/store/RedisStore.js#L24

So you should spot if the cause is in the `this.timeWindow` option or on another part of the system

**Comment by leorossi** [comment]
I have investigated on this a bit

@kruczyna that tests fails _sometimes_ . If you run `npx tap --grep="With redis store" test/route-rate-limit.test.js` you'll see that it always passes. This is though a separate issue.

Main problem here is that, when _not_ using `config` in your route (like the example provided in this issue) RedisStore is instanciated [here](https://github.com/fastify/fastify-rate-limit/blob/master/index.js#L75) during plugin registration, and the `globalParams.timeWindow` value is set to 60 seconds.

To fix this I can provide a PR where the `incr` function of the store takes a `timeWindow` parameter, and doesn't use the one used when the object was created.

**Comment by climba03003** [comment]
> 
> 
> I have investigated on this a bit
> 
> @kruczyna that tests fails _sometimes_ . If you run `npx tap --grep="With redis store" test/route-rate-limit.test.js` you'll see that it always passes. This is though a separate issue.
> 
> Main problem here is that, when _not_ using `config` in your route (like the example provided in this issue) RedisStore is instanciated [here](https://github.com/fastify/fastify-rate-limit/blob/master/index.js#L75) during plugin registration, and the `globalParams.timeWindow` value is set to 60 seconds.
> 
> To fix this I can provide a PR where the `incr` function of the store takes a `timeWindow` parameter, and doesn't use the one used when the object was created.

@leorossi I recommend to use the same logic inside `onRoute` handler. It should create a child `RedisStore` if it passed any custom options.
https://github.com/fastify/fastify-rate-limit/blob/master/index.js#L115-L118

**Comment by kruczyna** [comment]
Thank you guys for that, looking forward testing it out the PR!
In the meantime I'll allow myself to complete that unit test @Eomm was mentioning :)
